### PR TITLE
feat: add Analysis run to rollout  notifications

### PR DESCRIPTION
**Pull request by ashutosh16**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/) with a list of types and scopes found [here](https://github.com/argoproj/argo-rollouts/blob/master/.github/workflows/pr-title-check.yml), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ ] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

File: utils/annotations/annotations.go [hunk 1]
```diff
@@ -52,17 +53,30 @@ func GetWorkloadGenerationAnnotation(ro *v1alpha1.Rollout) (int32, bool) {
...
+		return 0, false
+	}
+	var obj interface{}
+
+	switch anyObj.(type) {
+	case *v1alpha1.Rollout:
+		obj = anyObj.(*v1alpha1.Rollout)
+	case *v1alpha1.AnalysisRun:
+		obj = anyObj.(*v1alpha1.AnalysisRun)
+	default:
```
Comment (zachaller) [inline utils/annotations/annotations.go:68]: Log a not supported type log

File: utils/record/record.go [hunk 2]
```diff
@@ -208,18 +210,12 @@ func (e *EventRecorderAdapter) defaultEventf(object runtime.Object, warn bool, o
...
 		}
 
-		apis, err := e.apiFactory.GetAPIsFromNamespace(namespace)
+		_, err := e.apiFactory.GetAPIsFromNamespace(namespace)
 		if err != nil {
 			logCtx.Errorf("notifications failed to get apis for eventReason %s with error: %s", opts.EventReason, err)
 			e.NotificationFailedCounter.WithLabelValues(namespace, name, opts.EventType, opts.EventReason).Inc()
 		}
 
-		for _, api := range apis {
```
Comment (zachaller) [inline utils/record/record.go:217]: We can't remove this

File: utils/annotations/annotations.go (hunk 1 above)
Comment (ashutosh16) [reply utils/annotations/annotations.go:67]: Done

File: utils/record/record.go [hunk 3]
```diff
@@ -42,9 +45,10 @@ func init() {
 }
 
 const (
-	controllerAgentName   = "rollouts-controller"
-	NotificationConfigMap = "argo-rollouts-notification-configmap"
-	NotificationSecret    = "argo-rollouts-notification-secret"
+	controllerAgentName     = "rollouts-controller"
+	NotificationConfigMap   = "argo-rollouts-notification-configmap"
+	NotificationSecret      = "argo-rollouts-notification-secret"
+	RolloutsPodTemplateHash = "rollouts-pod-template-hash"
```
Comment (zachaller) [inline utils/record/record.go:51]: Can we use the one defined in types.go

```
DefaultRolloutUniqueLabelKey string = "rollouts-pod-template-hash"
```

File: utils/record/record.go [hunk 4]
```diff
@@ -233,13 +238,77 @@ func (e *EventRecorderAdapter) K8sRecorder() record.EventRecorder {
...
+	sort.Slice(filteredArs[:], func(i, j int) bool {
+		ts1 := filteredArs[i].ObjectMeta.CreationTimestamp.Time
+		ts2 := filteredArs[j].ObjectMeta.CreationTimestamp.Time
+		return ts1.After(ts2)
+	})
+
+	var arsObj interface{}
+	arBytes, _ := json.Marshal(filteredArs)
+
+	if err != nil {
```
Comment (zachaller) [inline utils/record/record.go:271]: This error is not checked, can we return an error and add context to the error message

File: utils/record/record.go [hunk 5]
```diff
@@ -233,13 +238,77 @@ func (e *EventRecorderAdapter) K8sRecorder() record.EventRecorder {
...
+		}
+	}
+
+	sort.Slice(filteredArs[:], func(i, j int) bool {
+		ts1 := filteredArs[i].ObjectMeta.CreationTimestamp.Time
+		ts2 := filteredArs[j].ObjectMeta.CreationTimestamp.Time
+		return ts1.After(ts2)
+	})
+
+	var arsObj interface{}
```
Comment (zachaller) [inline utils/record/record.go:268]: can we switch `inerface{}` to `any`

File: utils/record/record.go [hunk 6]
```diff
@@ -233,13 +238,77 @@ func (e *EventRecorderAdapter) K8sRecorder() record.EventRecorder {
...
+				var vars = map[string]any{"rollout": obj, "time": timeExprs}
+
+				if arInformer == nil {
+					return vars
+				}
+
+				var ro v1alpha1.Rollout
+				err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj, &ro)
+
+				if err != nil || ro.Namespace == "" {
```
Comment (zachaller) [inline utils/record/record.go:298]: Whats the ro.Namespace="" about?

**Review by zachaller — APPROVED** [review]

**Pull request merged**